param_passing/proposal.md

```diff
@@ -0,0 +1,465 @@
+attribute `action`. The attribute takes in a string and the available actions
+are:
+
+* `add`: adds new elements to the original model
+* `modify`: modifies values and/or attributes of elements.
+This only updates existing elements and does not add or remove them.
+* `remove`: removes the elements from the original model
+* `replace`: replaces the elements from the original model to the new provided elements
+
+The `action` attribute must be provided either in the element identifier or in
+the direct children of the identifier.
+
+##### Alternatives considered
+
+Alternatively, instead of using the proposed element identifier, the user could
+use the attribute `id` for each customizable element in the original file but
+the intent of using the named path (i.e., the name of the parent elements and
+itself separated by double colons) was to provide a way for the user to modify
+the original model without requiring large amounts of updates to the original
+file.
+
+## Examples
+
+### Example 1
+
+Let's look at an example, here is an original model `base_robot`:
+
+```xml
+<model name="base_robot">
+  ...
+  <link name="chassis">
+    <visual name="lidar_visual">
+      ...
+    </visual>
+    <sensor name="lidar">
+      ...
+    </sensor>
+    <sensor name="camera">
+      <pose>0.5 0.02 0 0 0</pose>
+      ...
+    </sensor>
+  </link>
+
+  <link name="top">
+    <visual name="camera_visual">
+      ...
+      <geometry>
+        <box>
+          <size>0.03 0.03 0.05</size>
+        </box>
+      </geometry>
+      <material>
+        <ambient>0.5 0.5 0 1</ambient>
+        <diffuse>0 1 0 1</diffuse>
+        <specular>0.5 0.5 0.5 1</specular>
+      </material>
+    </visual>
+    <sensor name="camera">
+      <pose>0.2 1 0.5 0 0 0</pose>
+      <update_rate>20</update_rate>
+      ...
+    </sensor>
+  </link>
+```
+
+Now the user would like to use the original model `base_robot` in a world file
+but with modifications. The modifications include removing the lidar from the
+chassis link, adding a plugin to the camera sensor in the chassis link, adding a
+camera visual to the chassis link, update camera information in the top link,
+and replace the top link's camera visual geometry.
+
+```xml
+<sdf version="1.7"
+     xmlns:experimental="http://sdformat.org/schemas/experimental">
+  <model name="custom_robot">
+    <include>
+      <uri>model://base_robot</uri>
+
+      <experimental:params>
+        <visual element_id="chassis::lidar_visual" action="remove"/>
+        <sensor element_id="chassis::lidar" action="remove"/>
+
+        <plugin element_id="chassis::camera" action="add"
+                name="camera_depth_sensor" filename="libcamera_depth_sensor.so"/>
+
+        <visual element_id="chassis" name="camera_visual" action="add">
+          <pose>0.5 0.02 0 0 0</pose>
+          <geometry>
+              <box>
+                  <size>0.02 0.02 0.02</size>
+              </box>
```

Comment:
this block has different indentation from the rest. Use 2 spaces?

Reply:
d6aa791a54a569dc01bb88789b95a35a0f14fac2